src/uct/ib/mlx5/rc/rc_mlx5_common.c

```diff
@@ -608,63 +608,75 @@ uct_rc_mlx5_dp_ordering_ooo_init(uct_ib_mlx5_md_t *md,
                                  uct_rc_mlx5_iface_common_config_t *config,
                                  const char *tl_name)
 {
-    uct_ib_mlx5_dp_ordering_t min_forced_ordering;
+    static const char *dp_ordering_names[]    = {
+        [UCT_IB_MLX5_DP_ORDERING_IBTA]    = "IBTA",
+        [UCT_IB_MLX5_DP_ORDERING_OOO_RW]  = "OOO_RW",
+        [UCT_IB_MLX5_DP_ORDERING_OOO_ALL] = "OOO_ALL"
+    };
+    uct_ib_mlx5_dp_ordering_t max_dp_ordering = dp_ordering_cap;
+    uct_ib_mlx5_dp_ordering_t min_dp_ordering = UCT_IB_MLX5_DP_ORDERING_IBTA;
+    int force                                 = 0;
+    char ar_enable_str[16], ddp_enable_str[16];
 
     /*
      * HCA has an mlxreg admin configuration to force enable adaptive routing
      * (AR) or not.
      *
      * HCA cap/cap_2 booleans:
      * - if dp_ordering_ooo is set, QPC/DCTC can enable AR.
-     * - if UCT_IB_MLX5_MD_FLAG_DP_ORDERING_FORCE is set, QPC/DCTC can request mlxreg
-     *   configuration override, useful to force disable.
+     * - if UCT_IB_MLX5_MD_FLAG_DP_ORDERING_FORCE is set, QPC/DCTC can request
+     *   mlxreg configuration override, useful to force disable.
      */
-    iface->config.dp_ordering_force =
-            ucs_ternary_auto_value_is_yes_or_no(config->super.ar_enable) ||
-            ucs_ternary_auto_value_is_yes_or_no(config->ddp_enable);
-
-    /*
-     * Fail at the following cases:
-     * - Want to force configuration but can't force it
-     * - Want to force enable or force disable adaptive routing but it is unavailable
-     * - Want to force disable adaptive routing but force enabling DDP
-    */
-    if (iface->config.dp_ordering_force &&
-        !(md->flags & UCT_IB_MLX5_MD_FLAG_DP_ORDERING_FORCE)) {
-        goto failure;
-    }
 
     if (config->ddp_enable == UCS_YES) {
-        min_forced_ordering = UCT_IB_MLX5_DP_ORDERING_OOO_ALL;
-    } else if (config->super.ar_enable == UCS_YES) {
-        min_forced_ordering = UCT_IB_MLX5_DP_ORDERING_OOO_RW;
-    } else {
-        min_forced_ordering = UCT_IB_MLX5_DP_ORDERING_IBTA;
+        /* At least OOO_ALL is forced */
+        min_dp_ordering = ucs_max(min_dp_ordering,
+                                  UCT_IB_MLX5_DP_ORDERING_OOO_ALL);
```

Comment:
why do we need `ucs_max` here? can just assign to `UCT_IB_MLX5_DP_ORDERING_OOO_ALL`

Reply:
it's true that since we calculate min/max here right after they are initialized to the initial values, no need to really use min.max, but IMO this improves code readability and shows that applying DDP and AR configurations is interchangeable.